### PR TITLE
PipelineTesterMixin parameter configuration refactor

### PR DESCRIPTION
**Pull request by williamberman**
# Currently used test config

## `allowed_required_args` 
- testing that _all_ required parameters on _all_ pipelines are from a subset of variable names -> ends up not really testing much because when we add a new pipeline with a new required parameter, we just add it to the list. this isn't really checking an invariant or anything.
- also being used to configure what parameters to batch in another location -> this is really separate configuration as there's no real requirement that required parameters are batched
- This also gets a bit confusing as we don't have consistent treatment of required parameters and in many cases don't have any required parameters. I.e. `prompt` which you would think is the canonical required parameter is really technically optional in most pipelines as we allow overriding it with `prompt_embeddings`.

## `required_optional_params`
This is a good config but we could add a few more of the commonly included optional parameters

## `num_inference_steps_args`
This was a bad hack that I added for testing batching on the unclip pipelines with many different inference stages. 

# Refactor
`num_inference_steps_args` -> remove and pass as argument to test helper function

`required_optional_parameters` is good but we want to add a few more of the default optional parameters i.e. callback/callback_steps.

Really we want to separate `allowed_required_args` into two parts:
- A config for a set of parameters that are included in all (optional and required) paramers -> `params`
- A config for batched parameters -> `batch_params`

Both `params` and `batch_params` are generally determined by the type of pipeline -- i.e. text to image, image variation. They are configured on the specific pipeline subclass with a good error message explaining why and how to configure them when they are not set. Because we don't have strict interfaces (i.e. not all image variation pipelines take height and width) on pipelines nor families of pipelines, `params` are easily tweaked from provided defaults with operations on `frozenset`.

### CODE REVIEW

File: tests/pipeline_params.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        "width",
+        "guidance_scale",
+        "negative_prompt",
+        "prompt_embeds",
+        "negative_prompt_embeds",
+        "cross_attention_kwargs",
+    ]
+)
+
+TEXT_TO_IMAGE_BATCH_PARAMS = frozenset(["prompt", "negative_prompt"])
```
Comment (williamberman) [inline tests/pipeline_params.py:23]: Default param values for families of pipelines

File: tests/test_pipelines_common.py [hunk 2]
```diff
@@ -35,17 +35,22 @@ class PipelineTesterMixin:
...
+            "num_images_per_prompt",
+            "eta",
+            "generator",
+            "latents",
+            "output_type",
+            "return_dict",
+            "callback",
+            "callback_steps",
+        ]
+    )
```
Comment (williamberman) [inline tests/test_pipelines_common.py:53]: new set of `required_optional_params`

File: tests/test_pipelines_common.py [hunk 3]
```diff
@@ -76,6 +81,35 @@ def get_dummy_inputs(self, device, seed=0):
...
+            "`params` are checked if they can be passed to `__call__`. "
+            "`params` are used for families of pipelines that have common parameters such as text to "
+            "image pipelines for prompts and prompt embedding overrides. `pipeline_params.py` provides "
+            "some common sets of parameters such as `TEXT_TO_IMAGE_PARAMS`, `IMAGE_VARIATION_PARAMS`, etc... "
+            "If your pipeline's set of arguments has minor changes from one of the common sets of arguments, "
+            "do not make modifications to the existing common sets of arguments. I.e. a text to image pipeline "
+            "with non-configurable height and width arguments should set the attribute as "
+            "`params = TEXT_TO_IMAGE_PARAMS - {'height', 'width'}`. "
+            "See existing pipeline tests for reference."
+        )
```
Comment (williamberman) [inline tests/test_pipelines_common.py:97]: `params` with error message/docs

File: tests/test_pipelines_common.py [hunk 4]
```diff
@@ -76,6 +81,35 @@ def get_dummy_inputs(self, device, seed=0):
...
+            "You need to set the attribute `batch_params` in the child test class. "
+            "`batch_params` are the parameters required to be batched when passed to the pipeline's "
+            "`__call__` method. `pipeline_params.py` provides some common sets of parameters such as "
+            "`TEXT_TO_IMAGE_BATCH_PARAMS`, `IMAGE_VARIATION_BATCH_PARAMS`, etc... If your pipeline's "
+            "set of batch arguments has minor changes from one of the common sets of batch arguments, "
+            "do not make modifications to the existing common sets of batch arguments. I.e. a text to "
+            "image pipeline `negative_prompt` is not batched should set the attribute as "
+            "`batch_params = TEXT_TO_IMAGE_BATCH_PARAMS - {'negative_prompt'}`. "
+            "See existing pipeline tests for reference."
+        )
```
Comment (williamberman) [inline tests/test_pipelines_common.py:111]: batch_params with error message/docs

**Comment by williamberman** [comment]
note failing onnx tests seem unrelated

File: tests/test_pipelines_common.py (hunk 2 above)
Comment (sayakpaul) [reply tests/test_pipelines_common.py:53]: Love it.

File: tests/pipelines/dance_diffusion/test_dance_diffusion.py [hunk 5]
```diff
@@ -31,6 +32,16 @@
...
 class DanceDiffusionPipelineFastTests(PipelineTesterMixin, unittest.TestCase):
     pipeline_class = DanceDiffusionPipeline
+    params = UNCONDITIONAL_AUDIO_GENERATION_PARAMS
+    required_optional_params = PipelineTesterMixin.required_optional_params - {
+        "callback",
+        "latents",
+        "callback_steps",
+        "output_type",
+        "eta",
+        "num_images_per_prompt",
```
Comment (sayakpaul) [inline tests/pipelines/dance_diffusion/test_dance_diffusion.py:42]: I'd group them into a meaningful variable and then use it instead of defining ad-hoc.

**Review by sayakpaul — APPROVED** [review]

File: tests/test_pipelines_common.py (hunk 3 above)
Comment (yiyixuxu) [reply tests/test_pipelines_common.py:97]: error message is a little bit confusing to me - tried to edit it but also not 100% sure if I understood correctly

```suggestion
    @property
    def params(self) -> frozenset:
        raise NotImplementedError(
            "You need to set the attribute `params` in the child test class. "
            "`params` are used to check if all required parameters are present in `__call__`'s signature."
            " You can set `params` using one of the common set of parameters defined in`pipeline_params.py`"
            " e.g., `TEXT_TO_IMAGE_PARAMS` defines the common parameters used in text to  "
            "image pipelines, including prompts and prompt embedding overrides."
            "If your pipeline's set of arguments has minor changes from one of the common sets of arguments, "
            "do not make modifications to the existing common sets of arguments. I.e. a text to image pipeline "
            "with non-configurable height and width arguments should set the attribute as "
            "`params = TEXT_TO_IMAGE_PARAMS - {'height', 'width'}`. "
            "See existing pipeline tests for reference."
        )
```

**Review by yiyixuxu — APPROVED** [review]

File: tests/test_pipelines_common.py (hunk 3 above)
Comment (williamberman) [reply tests/test_pipelines_common.py]: nice! edits look good, made one small change https://github.com/huggingface/diffusers/pull/2502/commits/86ec20922d7c740c0e463bb1384b3b8513a3b5e2

File: tests/pipelines/dance_diffusion/test_dance_diffusion.py (hunk 5 above)
Comment (patrickvonplaten) [reply tests/pipelines/dance_diffusion/test_dance_diffusion.py:42]: yes this is a bit unreadable IMO, would be nicer to see directly what params are required

File: tests/test_pipelines_common.py [hunk 6]
```diff
@@ -28,17 +28,22 @@ class PipelineTesterMixin:
...
-    required_optional_params = ["generator", "num_inference_steps", "return_dict"]
-    num_inference_steps_args = ["num_inference_steps"]
+    # Canonical parameters that are passed to `__call__` regardless
+    # of the type of pipeline. They are always optional and have common
+    # sense default values.
+    required_optional_params = frozenset(
+        [
+            "num_inference_steps",
+            "num_images_per_prompt",
+            "eta",
```
Comment (patrickvonplaten) [inline tests/test_pipelines_common.py:38]:
```suggestion
```

I'd remove `"eta"` here. `"eta"` is a very specific case that doesn't apply to all pipelines. We probably shouldn't have added this at all in the first place

File: tests/test_pipelines_common.py (hunk 4 above)
Comment (patrickvonplaten) [reply tests/test_pipelines_common.py:104]: Very cool!

File: tests/pipelines/dance_diffusion/test_dance_diffusion.py (hunk 5 above)
Comment (patrickvonplaten) [reply tests/pipelines/dance_diffusion/test_dance_diffusion.py:42]: Actually ok for me as is - I'd just move "eta" out of the "required" optional base params. ETA is too specific & we shouldn't incentivize adding it in the future

File: tests/pipeline_params.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+CLASS_CONDITIONED_IMAGE_GENERATION_BATCH_PARAMS = frozenset(["class_labels"])
+
+UNCONDITIONAL_IMAGE_GENERATION_PARAMS = frozenset(["batch_size"])
+
+UNCONDITIONAL_IMAGE_GENERATION_BATCH_PARAMS = frozenset([])
+
+UNCONDITIONAL_AUDIO_GENERATION_PARAMS = frozenset(["batch_size"])
+
+UNCONDITIONAL_AUDIO_GENERATION_BATCH_PARAMS = frozenset([])
```
Comment (patrickvonplaten) [inline tests/pipeline_params.py:104]: Super nice! This is really helpful to classify the pipelines into certain tasks and should help us write new tests going forward

**Review by patrickvonplaten — APPROVED** [review]
Looks very nice! My only suggestion is to move "eta" out of the required/default list of optional parameters. ETA is only relevant for DDIM and is also generally not used. 

Going forward I actually don't think it's a good idea to have a "eta" argument

**Pull request merged**